### ISSUE TITLE
[S2-ALL-CP-DEPLOY] CP-DEPLOY — Deployment & Measurement Readiness Checkpoint

### ISSUE THREAD

**Issue by eid-abdelrihem** [microsoftproject26/Privacy-Preserving-Shopping-Intelligence#44]
| Field | Value |
|---|---|
| Task ID | `S2-ALL-CP-DEPLOY` |
| Sprint | Phase 1 — Sprint Week 2 |
| Owner | @eid-abdelrihem, @AhmedAbdelhamed01, @ahmedd-sherif |
| Lane | `ALL` |
| Priority | `P0` |
| Initial Board Status | `Backlog` |
| Estimated effort | 1.5 hours |
| Dependency wave | 13 |

## Goal

Confirm that ONNX export and system measurement tooling are reliable before final multi-task deployment evidence is produced.

The completed state is a validated, downstream-consumable set of: `docs/adr/ADR-003-deployment-measurement-readiness.md`.

## Why This Matters

Late discovery of broken parity or inconsistent benchmarks would invalidate system claims.

## Context

This is a Sprint Week 2 task in the `ALL` lane. It consumes [#38](https://github.com/eid-abdelrihem/Privacy-Preserving-Shopping-Intelligence/issues/38) (`S2-SE-01`), [#39](https://github.com/eid-abdelrihem/Privacy-Preserving-Shopping-Intelligence/issues/39) (`S2-SE-02`), [#40](https://github.com/eid-abdelrihem/Privacy-Preserving-Shopping-Intelligence/issues/40) (`S2-SE-03`), [#41](https://github.com/eid-abdelrihem/Privacy-Preserving-Shopping-Intelligence/issues/41) (`S2-SE-04`), [#50](https://github.com/eid-abdelrihem/Privacy-Preserving-Shopping-Intelligence/issues/50) (`S2-SE-05`), [#27](https://github.com/eid-abdelrihem/Privacy-Preserving-Shopping-Intelligence/issues/27) (`S1-SE-07`), produces `docs/adr/ADR-003-deployment-measurement-readiness.md`, and hands those outputs to [#53](https://github.com/eid-abdelrihem/Privacy-Preserving-Shopping-Intelligence/issues/53) (`S2-SE-08`).

The Issue owns only this producer contract. Engineering choices may vary inside the documented flexibility, but schema/protocol identities, hashes and downstream-visible behavior must remain stable once frozen.

## Inputs Required

- The authoritative V2 execution package and merged kickoff decisions in [#1](https://github.com/eid-abdelrihem/Privacy-Preserving-Shopping-Intelligence/issues/1) (`S1-ALL-01`).
- **Hard start:** [#38](https://github.com/eid-abdelrihem/Privacy-Preserving-Shopping-Intelligence/issues/38) (`S2-SE-01`) — consume `ONNX export module under package-root convention; initial model ONNX URI/SHA and manifest; PyTorch/ONNX parity report`. Real implementation cannot start before it is Done.
- **Hard start:** [#39](https://github.com/eid-abdelrihem/Privacy-Preserving-Shopping-Intelligence/issues/39) (`S2-SE-02`) — consume `INT8 ONNX URI/SHA; SystemMeasurement v1 FP32/INT8 benchmark; per-task quality-size-latency comparison`. Real implementation cannot start before it is Done.
- **Hard start:** [#40](https://github.com/eid-abdelrihem/Privacy-Preserving-Shopping-Intelligence/issues/40) (`S2-SE-03`) — consume `machine-readable collision statistics; collision/deployment evidence report with hashes`. Real implementation cannot start before it is Done.
- **Hard start:** [#41](https://github.com/eid-abdelrihem/Privacy-Preserving-Shopping-Intelligence/issues/41) (`S2-SE-04`) — consume `ONNX parity CI test; local parity test command`. Real implementation cannot start before it is Done.
- **Hard start:** [#50](https://github.com/eid-abdelrihem/Privacy-Preserving-Shopping-Intelligence/issues/50) (`S2-SE-05`) — consume `application-payload byte boundary and two-client oracle`. Real implementation cannot start before it is Done.
- **Hard start:** [#27](https://github.com/eid-abdelrihem/Privacy-Preserving-Shopping-Intelligence/issues/27) (`S1-SE-07`) — consume `SystemMeasurement v1 schema/spec; measurement utility under package-root convention; golden JSON and known-value tests`. Real implementation cannot start before it is Done.

### Hard Start Dependencies

- [#38](https://github.com/eid-abdelrihem/Privacy-Preserving-Shopping-Intelligence/issues/38) (`S2-SE-01`) — MUST be Done before real implementation starts.
- [#39](https://github.com/eid-abdelrihem/Privacy-Preserving-Shopping-Intelligence/issues/39) (`S2-SE-02`) — MUST be Done before real implementation starts.
- [#40](https://github.com/eid-abdelrihem/Privacy-Preserving-Shopping-Intelligence/issues/40) (`S2-SE-03`) — MUST be Done before real implementation starts.
- [#41](https://github.com/eid-abdelrihem/Privacy-Preserving-Shopping-Intelligence/issues/41) (`S2-SE-04`) — MUST be Done before real implementation starts.
- [#50](https://github.com/eid-abdelrihem/Privacy-Preserving-Shopping-Intelligence/issues/50) (`S2-SE-05`) — MUST be Done before real implementation starts.
- [#27](https://github.com/eid-abdelrihem/Privacy-Preserving-Shopping-Intelligence/issues/27) (`S1-SE-07`) — MUST be Done before real implementation starts.

### Close-Only Dependencies

- None.

### V2 work retained
1. Review ONNX export and PyTorch↔ONNX parity.
2. Review ONNX parity CI.
3. Review benchmark protocol for model size, peak RAM, CPU inference latency, and local-update latency where available.
4. Review quantization result and hash-collision analysis.
5. Review actual Flower upload/download instrumentation and byte-definition documentation.
6. Confirm no unresolved P0 deployment/measurement blocker remains and record GO/NO-GO.

### Task-specific technical requirements
1. Use an evidence matrix covering model/config hash, opset/I-O, parity fixtures/errors, benchmark device/provider/threads/warmups/repetitions/median/p95, FP32/INT8 size/quality, RSS, collision evidence and the byte oracle.
2. Record accepted/deferred findings and explicit GO/NO-GO.

## Implementation Notes / Recommended Approach

- Consume SystemMeasurement v1 records rather than manually copying benchmark values.
- Use the mature tools already selected by the owning lanes and make decisions from versioned evidence rather than rebuilding infrastructure.
- If V2 does not freeze a physical path, reuse an existing source/test/schema root. If none exists, the first owning PR selects one canonical repository-relative path, records the logical-name → path/URI mapping, and every consumer reuses it.

## Decision Required

The engineering work below must not silently select the following unresolved scientific or policy choices. Record the approved outcome in the named ADR/gate/config before the stated blocking point.

#### Decision

Freeze acceptable per-task quality degradation and quantitative CP-DEPLOY acceptance criteria.

#### Why It Cannot Be Chosen Locally

A size/latency win is not acceptable if it invalidates task quality or parity.

#### Candidate Options

- Evaluate the original-plan candidate: at least 2× smaller, no more than 1% relative loss on each approved headline metric, and not slower under the same benchmark
- Approve a different evidence-backed per-task bound before viewing final test results
- Reject/defer INT8 and retain FP32 when the approved rule is exceeded

#### Evidence Required

FP32/INT8 metrics on the same frozen manifest, size, latency, hardware metadata and operator compatibility.

#### Downstream Impact

S2-SE-02, CP-DEPLOY, S2-SE-08 and final evidence.

#### Approval Required From

All three members; System leads evidence and Data & Model approves metric impact.

#### Blocking Effect

Benchmark implementation may proceed; INT8 cannot be accepted as deployable evidence until approved.

#### Decision

Approve GO/NO-GO using parity, benchmark, collision, quantization and byte-definition evidence.

#### Why It Cannot Be Chosen Locally

Final ONNX/measurement evidence must not silently waive a P0 failure.

#### Candidate Options

- GO with all mandatory criteria and approved INT8 disposition
- NO-GO with exact blocker; FP32 may remain evidence only when explicitly approved

#### Evidence Required

Model/opset/I-O hashes, parity errors, benchmark protocol, FP32/INT8 deltas, RSS and byte oracle.

#### Downstream Impact

CP-DEPLOY ADR and S2-SE-08.

#### Approval Required From

All three members; System leads and Data & Model reviews quality.

#### Blocking Effect

Review can proceed; final multi-task deployment pass is blocked on NO-GO.

## Allowed Flexibility

- ADR organization and extra diagnostics.
- Add non-breaking diagnostics, documentation and performance improvements that preserve all versioned inputs, outputs, metrics and hashes.

## Requires Team Approval

- Quantitative acceptance, INT8 disposition and GO/NO-GO.
- The unresolved **INT8 acceptance** decision described above.
- The unresolved **Deployment readiness GO/NO-GO** decision described above.
- Any change to a frozen cross-lane contract or to this Issue's hard-start/close-only semantics.
- If one is necessary: stop the affected portion, preserve evidence, and request the named team decision/ADR; do not select a convenient default.

## Do Not Do

- Do not compare unlike benchmark protocols, hide a failed head or estimate bytes.
- Merge only after the required validation passes.

## Expected Deliverables

- `docs/adr/ADR-003-deployment-measurement-readiness.md`

Artifacts without a V2-frozen physical path must use the one-time path-selection convention above and be registered through `docs/artifacts.md`; do not create parallel aliases.

## Handoff — Who Needs This?

- **Producer `S2-ALL-CP-DEPLOY` → `docs/adr/ADR-003-deployment-measurement-readiness.md` → consumer [#53](https://github.com/eid-abdelrihem/Privacy-Preserving-Shopping-Intelligence/issues/53) (`S2-SE-08`)** (SYSTEM_DEPLOYMENT_SECURITY / @ahmedd-sherif). The consumer uses this interface for **Multi-Task ONNX Export, Parity & Quantization Pass**.

## Validation / Tests

- All hashes/units resolve; known-bad parity fails; benchmark reproduces; byte oracle and approved INT8 disposition are present.
- The PR records one exact local validation command and its expected pass/fail outcome; CI uses the same command where the task is CI-safe.
- A schema/version/hash mismatch must fail visibly rather than be silently coerced.

## Edge Cases / Failure Modes

- Unsupported operator, numerical drift, warmup bias, per-head INT8 degradation and missing measurement.

## Reproducibility

- Record runtime/quantizer/provider/hardware/config identities.
- Record the exact entry point/command, output logical ID/path/URI and SHA-256 for every persisted artifact.

## Definition of Done

- [ ] Readiness decision is GO.
- [ ] Parity and benchmark protocols are validated.
- [ ] No unresolved P0 deployment/measurement issue remains.
- [ ] Every expected deliverable exists at its single canonical path/URI and its schema/config/protocol version and SHA-256 are recorded where applicable.
- [ ] All task-specific validation oracles and known-bad fixtures pass/fail exactly as documented.
- [ ] The producer-to-consumer contract is validated with the real downstream interface.
- [ ] Every Decision Required above is approved by the named reviewers and recorded before the blocking freeze/closure point.
- [ ] Implementation/evidence is submitted through a PR and required validation passes.
- [ ] The PR is merged; only then may the Project Status move to Done.

## If Blocked

1. Identify the exact missing artifact, decision, interface, permission or failing invariant and preserve the error/evidence.
2. Comment on this Issue with what is missing, the responsible Issue/person, completed work and the smallest reproducible failure.
3. Move to `Blocked` only when this task was actively started and no safe in-scope work remains.
4. Do not invent a workaround that changes a frozen scientific or cross-lane contract.
5. If possible, move to the next dependency-cleared Ready Issue in the `ALL` lane; a blocked task must not idle the whole lane.

## Next Recommended Work

- Select the next dependency-cleared Ready task in the `ALL` lane; do not start another owner's downstream task merely because this handoff is complete.

## Workflow

Backlog → Ready → In Progress → In Review → Done. Use Blocked only for a real blocker encountered after work starts. Normal WIP is one active implementation Issue per member; a safe unattended machine job permits one small independent Ready task.

**Comment by ahmedd-sherif** [comment]
Two of this checkpoint's three open blockers are now closed:

- #40 product hash collision report: merged, evidence on main.
- #50 Flower upload/download byte measurement: PR #83 open.

#38 remains open and still needs #30.

One thing in this issue's dependency list needs a decision rather than an
update. It still lists #39 and #41 as Hard Start Dependencies, but both were
closed as not planned and folded elsewhere: #39's INT8 attempt moved into #53,
and #41's ONNX parity moved into #38.

#53 is listed as a consumer of this checkpoint's ADR. So as written, the INT8
acceptance decision here depends on evidence produced by a task that is
supposed to consume the decision. That circle cannot close as stated.

The straightforward resolution is to drop Decision 1 from this checkpoint and
let #53 make the FP32-or-INT8 call against the acceptance rule already written
here, with this checkpoint keeping only the deployment readiness GO/NO-GO.
Either way the dependency list should stop naming two closed issues.

**Comment by ahmedd-sherif** [comment]
Update: every hard dependency listed on this checkpoint is now closed.

| Dependency | State |
|---|---|
| #38 S2-SE-01 ONNX export and parity | Closed |
| #39 S2-SE-02 INT8 | Closed as not planned, folded into #53 |
| #40 S2-SE-03 hash collisions | Closed |
| #41 S2-SE-04 ONNX parity in CI | Closed as not planned, folded into #38 |
| #50 S2-SE-05 Flower byte measurement | Closed |
| #27 S1-SE-07 runtime measurements | Closed |

So this checkpoint is startable. The circularity I raised earlier is resolved in
the only way that does not require rewriting the graph: rather than wait for
#53 to produce INT8 evidence that #53 is supposed to consume this decision for,
I produced the evidence now against the current model.

## INT8 evidence for Decision 1

`docs/deployment-int8-benchmark.md` and
`docs/evidence/s2-se-02/int8_benchmark.v1.json`.

Against the three conditions this checkpoint states:

| Condition | Result |
|---|---|
| At least 2x smaller | Met. 14,678,873 to 4,003,834 bytes, **3.67x** |
| Not slower | Met at the median. 1.22 ms to 1.14 ms, one thread, 30 repetitions |
| No more than 1% relative loss per headline metric | **Not measured** |

The third one is the one to read carefully. It asks for a relative loss on
MRR@20, PR-AUC and NDCG@5 computed on the frozen evaluation set. That set
carries `user_id` and is not in this repository, so this run cannot produce it.

What I can bound is how far the outputs move: 0.0908 on `t1_logits`, 0.0765 on
`t2_logit`, 0.0000 on `t3_scores`. A shift of about 0.09 on an unnormalised
logit is not negligible, and whether it survives into the ranking metrics is
exactly the open question. The T3 zero is not reassurance either, because the
T3 path is dominated by the frozen retrieval rank, which quantization leaves
alone.

Two further caveats worth recording in the ADR rather than discovering later.
The p95 goes the other way, 2.28 ms FP32 against 2.90 ms INT8, though on a
one-millisecond forward pass that is scheduling noise rather than the model. And
these numbers come from a seed-built model, not a trained one; size and latency
carry over because the operator mix is identical, but quantization error depends
on the weight distribution, so the output drift does not carry over with the
same confidence.

## What I would put in the ADR

Accept INT8 on size and speed, record quality as unanswered, and require #53 to
recompute the per-task metric delta against the frozen evaluation set before
INT8 becomes the shipped artifact, keeping FP32 if any delta exceeds 1%.

Also worth doing while we are here: drop #39 and #41 from the dependency list
above. Both are closed as not planned and naming them suggests work that will
never arrive.

**Comment by ahmedd-sherif** [comment]
`docs/adr/ADR-003-deployment-measurement-readiness.md` is on main, merged in PR
#95. That is this checkpoint's declared deliverable, and every hard dependency
it lists is closed.

**Decision 1 — Quantization: FP32.** INT8 is 3.51x smaller and 26% slower on the
approved candidate. One of the three stated conditions fails outright, so the
rule rejects it. The prototype benchmark pointed the other way and both runs are
kept in the record; dynamic quantization pays a per-call cost that scales with
history length and candidate count, so the prototype's number would have led to
the wrong call.

**Decision 2 — Readiness: GO**, with three conditions written into the ADR rather
than left implicit: the committed evidence was produced without trained weights
so it measures the shipped architecture and not the shipped model; the INT8
quality delta is unanswered rather than passed; and no privacy claim rests on the
document.

Status is Proposed pending approval from all three of us, which is what this
checkpoint requires. @eid-abdelrihem @AhmedAbdelhamed01 — if you agree, say so
here and I will flip the status line to Accepted and close this.

If you disagree with the FP32 call, the thing to push back on is the latency
measurement: one thread, five warmups discarded, thirty repetitions, same
machine and shapes for both models. Reproduce with
`uv run --locked python scripts/deployment/export_final_model.py --report <path>`.

One housekeeping item either way: this issue lists #39 and #41 as hard start
dependencies. Both were closed as not planned and folded into #53 and #38. Their
evidence exists and is cited in the ADR, but the dependency list should drop the
two names so it stops implying work that will never arrive under those numbers.
